feat: Snowflake adapter using show statements

derekbunch opened the pull request:
rebased https://github.com/kndndrj/nvim-dbee/pull/109 on master and updated it to use show commands to avoid waking warehouses to prevent extra cost

im not a go dev, but tried my best to follow the patterns i saw as much as possible

![image](https://github.com/user-attachments/assets/2927c699-daf1-4c38-b568-2ee18cbfef3c)

derekbunch committed:
snowflake branch base copy

derekbunch committed:
updateing to use show instead of select, not working yet

derekbunch committed:
columns and structure

derekbunch committed:
show transient databases

derekbunch committed:
go mod tidy

derekbunch commented:
i couldn't figure out a way to do `show databases` and get the current without 2 queries and that seemed to cause issues, so I left that one in there as I assume it happens much less frequently than structure and columns. 

If anybody knows how i could do that, happy to add it.

MattiasMTS commented:
Hi @derekbunch sorry for slow response, feel free to ping me for faster help :) 

Is this ready for review? We are currently focusing on adding somewhat integration tests for existing adapters. Ideally, any new adapter should also try to follow this structure.

We are using [testcontainers](https://golang.testcontainers.org/) as the tool for testing each adapter. From quickly checking, it seem there is a localstack image https://hub.docker.com/r/localstack/snowflake that you can use to emulate the snowflake warehouse locally. Could you try and take a stab adding this into this PR as well? It would be very much appreciated as it simplifies testing and debugging quite a lot. Let me know if you need any help with this!

derekbunch commented:
Hi @MattiasMTS, yes it should be ready for review.

And no problem. ill look into the links you shared and try to get those tests implemented for this MR.

MattiasMTS commented on dbee/adapters/snowflake_driver.go:
```diff
@@ -0,0 +1,163 @@
+	url.URL
+}
+
+func (c *SnowflakeURL) String() string {
```
edit: ignore this, see comment about the URL prefix

---

~~actually, can't this be simplified by doing:~~
```golang
func (s *snowflakeURL) String() string {
	dsn := s.URL.String()
	return strings.TrimPrefix(dsn, "snowflake://")
}
```
~~and also add some unit test for this wrapping logic:~~

```golang
func TestSnowflakeURL_String(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		want string
	}{
		{
			name:     "should return dsn without snowflake prefix",
			input:    "snowflake://user:pass@myaccount/mydb",
			want: "user:pass@myaccount/mydb",
		},
		{
			name:     "should return input string if snowflake prefix not present",
			input:    "user:pass@myaccount/mydb?warehouse=compute",
			want: "user:pass@myaccount/mydb?warehouse=compute",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			parsed, err := url.Parse(tt.input)
			require.NoError(t, err)

			surl := snowflakeURL{parsed}
			got := surl.String()
			require.Equal(t, tt.want, got)
		})
	}
}
```

MattiasMTS commented on dbee/adapters/snowflake_test.go:
```diff
@@ -0,0 +1,48 @@
+				r.Error(err)
+				return
+			}
+			r.NoError(err)
```
the `want` here is never checked. Missing `require.Equal(t, tt.want, got)`, no need to use a `require.New` here :)

**edit**: feel free to ignore this if you're going to just add the `String` method :)

MattiasMTS commented on dbee/adapters/snowflake_test.go:
```diff
@@ -0,0 +1,48 @@
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := new(Snowflake).Connect(tt.args.rawURL)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("NewSnowflake() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
```
I think this can be skipped, or what are you trying to check here?
**edit**: feel free to ignore this if you're going to just add the `String` method :)

MattiasMTS commented on dbee/adapters/snowflake_test.go:
```diff
@@ -0,0 +1,48 @@
+		wantErr bool
+	}{
+		{
+			name: "Valid URL",
```
**edit:** ignore if we decide to not use prefix in URL
~~feel free to take a look at the example test I gave to the `String` part - I believe that is the only unit test we need here given this `Connect` will be tested in integration test. So, let's just add a unit test for the `String` logic - or wdyt?~~

MattiasMTS reviewed (commented):
Nice work! Thanks a lot for the contribution!! ✨
I left some comments - first that should be addressed is the `URL` part of being able to exclude the `snowflake://` protocol in the URL as we are already declaring which driver via the `type` attribute 😋

MattiasMTS commented on dbee/adapters/snowflake_driver.go:
```diff
@@ -0,0 +1,163 @@
+// gosnowflake does not support the snowflake:// scheme
+// , it expects the full connection string excluding the scheme.
+// e.g. "snowflake://user:password@account/db" -> "user:password@account/db"
+type SnowflakeURL struct {
```
I realize this that I should try and focus on adding some hosted document related to this so it is clear to the users how each adapter works and what underlying driver it is using. Another thing on the TODO list! ✏️

derekbunch commented on dbee/adapters/snowflake_driver.go:
```diff
@@ -0,0 +1,163 @@
+			DATABASE_NAME AS database_name
+		FROM INFORMATION_SCHEMA.databases
+    WHERE DATABASE_NAME != CURRENT_DATABASE();
+  `
```
i was actually able to get around this by switching to using a snowflake config, which is alot easier to work with

derekbunch commented:
> Hi @derekbunch sorry for slow response, feel free to ping me for faster help :)
> 
> Is this ready for review? We are currently focusing on adding somewhat integration tests for existing adapters. Ideally, any new adapter should also try to follow this structure.
> 
> We are using [testcontainers](https://golang.testcontainers.org/) as the tool for testing each adapter. From quickly checking, it seem there is a localstack image https://hub.docker.com/r/localstack/snowflake (edit: and testcontainer: https://golang.testcontainers.org/modules/localstack/) that you can use to emulate the snowflake warehouse locally. Could you try and take a stab adding this into this PR as well? It would be very much appreciated as it simplifies testing and debugging quite a lot. Let me know if you need any help with this!

@MattiasMTS I've looked into this a bit and tried to get it set up. Looks like it needs an auth token to run the container so I created an account, generated an auth token, and tried to run the container with it and it gives me this error
```
❯docker run \
    --rm -it \
    -p 127.0.0.1:4566:4566 \
    -p 127.0.0.1:4510-4559:4510-4559 \
    -p 127.0.0.1:443:443 \
    -e LOCALSTACK_AUTH_TOKEN="<redacted>" \
    localstack/snowflake
                                                                                                                                                        
LocalStack version: 4.2.1.dev76
LocalStack build date: 2025-03-20
LocalStack build git hash: 6b2781950

Docker not available
2025-03-22T05:51:45.089  INFO --- [  MainThread] l.p.c.b.licensingv2        : Successfully requested and activated new license <redacted>:hobby 🔑✅
2025-03-22T05:51:47.017  INFO --- [  MainThread] l.p.c.extensions.platform  : loaded 1 extensions
⚠ The Snowflake emulator is currently not covered by your license. ❄ Please reach out to LocalStack Support to get access to this product. ⚠
```

Any idea how i should proceed? Do y'all use a ci token for your ci where this would run or is this a non-starter?

derekbunch committed:
address PR comments

MattiasMTS commented:
> Any idea how i should proceed? Do y'all use a ci token for your ci where this would run or is this a non-starter?

ahh, right. I guess that version of localstack requires a subscription.. That is a bit unfortunate. Feel free to skip this part/scope of this PR then and I'll try and find if we can get a sub based on OSS stuff. Thanks a lot for trying it out and feedback the info! 😄

MattiasMTS commented on dbee/adapters/snowflake.go:
```diff
@@ -0,0 +1,54 @@
+	}
+	connector := gosnowflake.NewConnector(gosnowflake.SnowflakeDriver{}, *config)
+	db := sql.OpenDB(connector)
+	err = db.Ping()
+	if err != nil {
+		return nil, err
+	}
```
let's ping with a timeout here, e.g

```go
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	if err := db.PingContext(ctx); err != nil {
		return nil, fmt.Errorf("unable to ping snowflake: %w", err)
	}
```

if you want to ping it

derekbunch replied:
😮 super cool, thanks! didnt know this is how you did that!

MattiasMTS commented on dbee/adapters/snowflake_driver.go:
```diff
@@ -0,0 +1,129 @@
+	config.Database = name
+	connector := gosnowflake.NewConnector(gosnowflake.SnowflakeDriver{}, config)
+	db := sql.OpenDB(connector)
+	err := db.Ping()
```
same comment here on ping as in the connector 😋

MattiasMTS commented on dbee/adapters/snowflake_test.go:
```diff
@@ -0,0 +1,48 @@
+	"github.com/kndndrj/nvim-dbee/dbee/core/builders"
+)
+
+func TestNewSnowflake(t *testing.T) {
```
I think you can remove this test since I am not sure we need it for testing anymore?

derekbunch replied:
ah yes, was just waiting to delete this until i confirmed how to handle testing!

MattiasMTS commented on dbee/adapters/snowflake_driver.go:
```diff
@@ -0,0 +1,129 @@
+// snowflakeDriver is a sql client for snowflakeDriver.
+type snowflakeDriver struct {
+	c      *builders.Client
+	config *gosnowflake.Config
```
feel free to make this a non-pointer if you are only using it in `SelectDatabase` with dereference, or wdyt?

MattiasMTS commented on dbee/adapters/snowflake.go:
```diff
@@ -0,0 +1,54 @@
+// or
+// host:port/database/schema?account=user_account[?param1=value1&paramN=valueN]
+// https://github.com/snowflakedb/gosnowflake/blob/b034584aa6fc171c1fa02e5af1f98234f24538fe/dsn.go#L308-#L314
+func (r *Snowflake) Connect(rawURL string) (core.Driver, error) {
```
looks great! We can iteratore more on what kind of connection inputs a user can have here later on 😄 This looks a lot cleaner now IMO! ✨

derekbunch replied:
yea, im also planning to update this (if needed) to support oauth once snowflake merges [this PR](https://github.com/snowflakedb/gosnowflake/compare/master...SNOW-1825500-oauth2) since that or 2FA is snowflake's preferred method of auth. Password based auth [will be disabled soon](https://medium.com/snowflake-engineering/important-snowflake-will-require-mfa-by-april-2025-or-your-access-will-be-blocked-bd999ca7ab27)

derekbunch committed:
address PR comments

phdah commented on dbee/adapters/snowflake_driver.go:
```diff
@@ -0,0 +1,131 @@
+		if err != nil {
+			return "", nil, err
+		}
+		databaseName := row[2].(string)
```
```suggestion
		databaseName := row[1].(string)
```
Database name is in the second column.

derekbunch committed:
fix database name column

derekbunch committed:
update snowflake and make sure oauth works

derekbunch committed:
expand home in private key path

derekbunch committed:
cache structure to fix ui hanging

derekbunch commented:
hey @MattiasMTS sorry for the dalay on this. i rebased this and updated it to work with the upstream snowflake oauth (since snowflake is deprecating password only access) and validated that everything works again. 

would love another look! 🙏🏻
